specifications/mempool/README.md

```diff
@@ -0,0 +1,280 @@
+```
+
+Expected input:
+
+* `txn` - transaction submitted by JSON RPC client proxied to mempool by JSON RPC service
+* `callback` - callback to JSON RPC service
+
+Required behavior:
+
+* validate and submit transaction to mempool
+* send back transaction submission result to callback
+
+```rust
+fn broadcast_single_peer(peer, peer_sync_data)
+```
+
+Expected input:
+
+* `peer` - peer to broadcast to
+* `peer_sync_data` - peer-specific information to determine which transactions to broadcast to this peer. This can contain information like the transaction broadcast history for this peer
+
+required:
+
+* crafts `MempoolSyncMsg::BroadcastTransactionsRequest` with broadcast-ready transactions for this peer
+* network-sends MempoolSyncMsg to peer
+
+```rust
+fn process_transaction_broadcast(msg: MempoolSyncMsg)
+```
+
+Expected input:
+
+* `MempoolSyncMsg::BroadcastTransactionsRequest`
+
+Required behavior:
+
+* validate and submit transactions to mempool
+* craft corresponding `MempoolSyncMsg::BroadcastTransactionsResponse` and send to peer (via `network::send_to(peer, msg)`)
+
+## Abstracted Modules
+
+### VM
+
+Mempool performs transaction signature verification and validations via the VM module. 
+
+```rust
+fn validate_transaction(&self, txn: SignedTransaction)
+```
+
+* [validates](#transaction-states-in-mempool) a transaction
+
+### Consensus
+
+Consensus interacts with mempool by (1) requesting a block of transactions
+ready for consensus, and (2) notifying mempool of rejected committed transactions
+- see [here](#functionality:-processing-consensus-requests) for the required 
+behavior for this interaction. Otherwise, implementing and coordinating this 
+interaction is an implementation detail.
+
+### State Sync
+
+Mempool receives notification from state sync about committed transactions -
+see [here](#functionality:-processing-state-sync-requests) for the required
+behavior for this interaction. Otherwise, implementing and coordinating 
+this interaction is an implementation detail.
+
+### Network
+
+Mempool interacts with the networking module to send messages
+defined in the [network protocol](#network-protocol) to peers in the network. 
+
+The following is a _sample_ interface that is not strictly required by the
+specification.
+
+```rust
+fn send_to(recipient: PeerId, message: MempoolSyncMsg)
+```
+
+* sends `message` to `recipient`
+
+### JSON RPC
+
+JSON RPC service proxies the transaction submission it receives to mempool with
+
+1. transaction to submit
+2. callback to send a transaction submission result to JSON RPC service
+
+## Network Protocol
+
+### MempoolSyncMsg
+
+Network message used for broadcast of pending transactions to remote peers
+
+```rust
+pub enum MempoolSyncMsg {
+    BroadcastTransactionsRequest {
+        request_id: Vec<u8>,
+        transactions: Vec<SignedTransaction>,
+    },
+    BroadcastTransactionsResponse {
+        request_id: Vec<u8>,
+        retry: bool,
+        backoff: bool,
+    },
+}
+```
```

Comment:
I think from a spec perspective this is all you really need

the rust code blocks kinda get into the nuances of implementation

Reply:
i think there are some extra descriptions that belong in spec though... e.g. request_id should be equal for pairs of requests/responses, etc